Fix AggregationFactories.Builder equality and hash regarding order

hendrikmuhs opened the pull request:
Fixes the equals and hash function to ignore the order of aggregations to ensure equality after serialization and deserialization.

Note:

At time of writing the impact of this issue mainly affects tests (-> non-issue), however in future work I am storing aggregation configuration as part of job configurations and therefore require this fix.

It also fixes a potential issue in caching when the same query contains aggregations in different order.  1st it will not hit in the cache, 2nd cache objects which shall be equal might end up twice in the cache.

Previous fix: https://github.com/elastic/elasticsearch/pull/33942

switches to a LinkedHashSet instead of a List.

CC @jimczi @cbuescher @colings86

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/AggregationBuilder.java:
```diff
@@ -79,12 +79,12 @@ public String getName() {
 
     /** Return the configured set of subaggregations **/
-    public List<AggregationBuilder> getSubAggregations() {
+    public Set<AggregationBuilder> getSubAggregations() {
```
Why not Collection ?

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/PipelineAggregationBuilder.java:
```diff
@@ -68,8 +68,8 @@ public String getName() {
      */
-    protected abstract void validate(AggregatorFactory<?> parent, List<AggregationBuilder> factories,
-            List<PipelineAggregationBuilder> pipelineAggregatorFactories);
+    protected abstract void validate(AggregatorFactory<?> parent, Set<AggregationBuilder> aggregationBuilders2,
```
Same here, `Collection` ?

jimczi commented on server/src/test/java/org/elasticsearch/search/aggregations/AggregatorFactoriesBuilderTests.java:
```diff
@@ -0,0 +1,162 @@
+/*
```
I think we have `BaseAggregationTestCase` for aggregators so we could maybe add the testing for this case directly there. It would test all implementations rather than just a few.

hendrikmuhs commented:
Thanks @jimczi! I will work on the details once we agreed on the approach.

hendrikmuhs commented:
Addressed the review comments. @jimczi @colings86 

The remaining test failure is another ordering problem: Prior to that change ordering was clearly defined by the order of aggregations in the query. I changed the IT to access by key rather than by position. 

For the doc/yaml build however: I do not know how to do this and wonder if this PR is good or not. There are potentially more tests that purely work on text matching. Order is not defined in JSON, still users might rely on it. This change makes the order somewhat unpredictable for a user. I think it would be nice to bring back the original behavior or always order in a defined way (lexical order).

Thoughts? Comments?

jimczi commented:
> For the doc/yaml build however: I do not know how to do this and wonder if this PR is good or not. There are potentially more tests that purely work on text matching. Order is not defined in JSON, still users might rely on order. This change makes the order somewhat unpredictable for a user. I think it would be nice to bring back the original behavior or always order in a defined way (lexical order).

I don't think we should do anything here. The order will change if you have a proxy or a client that processes the json differently so we should not force anything. If we have tests that rely on ordering we should change them but I am 99% sure ;) that the doc/yaml tests don't care about the order.

hendrikmuhs commented:
I take my 1% chance. ;-) There were 2 ordering problems in the doc tests, 1 was fixable the other one not. 

But, after changing to a `LinkedHashSet` from the ordinary `HashSet` I was able to reintroduce the old behavior of respecting the insertion order. This makes it more user friendly and fixes the doc test issues.

hendrikmuhs commented:
>  I don't think we need AggregatorFactoriesBuilderTests though. Instead we could change BaseAggregationTestCase to explicitly test serialization/equals/hashCode with more than one builder ?

This does not tackle the problem. The `BaseAggregationTestCase` is a base class for testing aggregations and it already tests `equals` and `hashCode`, all aggregations are fine w.r.t. to serialization. `AggregationBuilder` is fine as well.

The problem found is only in `AggregationFactories.Builder` which missed test coverage (`AggregationFactories` has tests).

I noted that the title is misleading, I will fix that.

jimczi commented:
> I take my 1% chance. ;-) There were 2 ordering problems in the doc tests, 1 was fixable the other one not.

I think it's fine to use a `LinkedHashSet` but the ordering issues are due to other APIs (`profile`) that use an array to reference the sub-aggregations. `AbstractAggregationBuilder` serializes the sub-aggregations as maps because that's what they really are:
````
{
	"size": 0,
	"aggs": {
		"1": {
			"terms": {
				"field": "country.keyword"
			},
			"aggs": {
				"2": {
					"avg": {
						"field": "value"
					}
				},
				"4": {
					"sum": {
						"field": "value"
					}
				}
			}
		}
	}
}
````
The fact that we use an array to represent them internally should not change the reasoning. The order of sub-aggregations in the request should not matter. 

> This does not tackle the problem. The BaseAggregationTestCase is a base class for testing aggregations and it already tests equals and hashCode, all aggregations are fine w.r.t. to serialization. AggregationBuilder is fine as well.

This pr fixes a real issue, today different orderings of the same sub-aggregations are cached individually. So this query for instance:
````
{
	"size": 0,
	"aggs": {
		"1": {
			"terms": {
				"field": "country.keyword"
			},
			"aggs": {
				"4": {
					"sum": {
						"field": "value"
					}
				},
				"2": {
					"avg": {
						"field": "value"
					}
				}
			}
		}
	}
}
````
... which yields the same response than the first example is considered as a different query by our cache. IMO this is a bug in the `AggregationBuilder` class which is why I think it should test this case in `BaseAggregationTestCase`. `equals` and `hashCode` are tested only when there is no or a single sub-agg which is why we never caught this issue before.

hendrikmuhs commented:
I now better understand what you mean and I can try to improve testcoverage of `BaseAggregationTestCase`. I do not think it can replace `AggregatorFactoriesBuilderTests` completely, because it is a an implementation of `AbstractSerializingTestCase` which ensures that `ToXContent` and `Writable` are correctly implemented. I think we should keep it.

`testUnorderedEqualsSubSet` - is added sugar, I happily move that.
